lizard_languages/clike.py

```diff
@@ -72,16 +72,85 @@ def _typedef(self, _, tokens):
         self.context.add_condition(-tokens.count("&&"))
 
 
+# pylint: disable=R0903
 class CLikeNestingStackStates(CodeStateMachine):
-    # pylint: disable=R0903
+    """Machinery to track nesting levels of tokens.
+
+    The main indicators of nesting are braces: '{' and '}'.
+    However, the code is complicated due to braceless structures,
+    which are control-flow structures in C-like languages.
+    Moreover, using braces as nesting level indicators leads to a big caveat;
+    C++ initializer lists, uniform initialization,
+    and any other constructs with braces add extra nesting level.
+
+    Another complication comes from nested classes inside function bodies
+    and control-flow declarations/bodies.
+    The handling of these complex cases is unspecified and can be ignored.
+    """
+
+    # Beasts that can be defined within one line without braces.
+    __braceless_structures = set(['if', 'else', 'for', 'while', 'do',
+                                  'switch'])
+    __paren_count = 0  # Used only to tackle the beasts.
+    __braceless = None  # Applies only to the beasts.
+    __structure_brace_stack = []  # Boolean stack for structures' brace states.
+
+    def __pop_braceless_structures(self):
+        """Pops structures up to the one with braces."""
+        self.context.pop_nesting()
+        is_structure = None
+        if self.__structure_brace_stack:
+            is_structure = self.__structure_brace_stack.pop()
+
+        while (is_structure is not None and self.__structure_brace_stack and
+                self.__structure_brace_stack[-1]):
+            self.__structure_brace_stack.pop()
+            self.context.pop_nesting()
+
+    def __else_if_structure(self, token):
+        """Handles possible compound 'else if' after 'else' token."""
+        self._state = self.__declare_structure
+        if token != "if":
+            self._state(token)
+
+    def __declare_structure(self, token):
+        """Ignores structures between parentheses on structure declaration."""
+        if token == "(":
+            self.__paren_count += 1
+        elif token == ")":
+            assert self.__paren_count > 0
```

Comment:
I agree with your point that errors shouldn’t be ignored silently and it should fail when there’s situation we cannot predict, and hopefully the user will tell us their findings (which happened quite often in the past).

But on the other hand I also hope lizard can be tolerable to broken/half-done code. And by “tolerable” I didn’t mean the error should be swallowed silently but be handled in a proper way.

Let me try to get back to the person who reported this assertion to me and see if I can get the code that triggered the assertion.

Thanks for reminding me this important design principle:-)

> On 5 Jun 2016, at 10:50 AM, Olzhas Rakhimov <notifications@github.com> wrote:
> 
> In lizard_languages/clike.py <https://github.com/terryyin/lizard/pull/123#discussion_r65810575>:
> 
> > +                self.__structure_brace_stack[-1]):
> > +            self.__structure_brace_stack.pop()
> > +            self.context.pop_nesting()
> > +
> > +    def __else_if_structure(self, token):
> > +        """Handles possible compound 'else if' after 'else' token."""
> > +        self._state = self.__declare_structure
> > +        if token != "if":
> > +            self._state(token)
> > +
> > +    def __declare_structure(self, token):
> > +        """Ignores structures between parentheses on structure declaration."""
> > +        if token == "(":
> > +            self.__paren_count += 1
> > +        elif token == ")":
> > +            assert self.__paren_count > 0
> I am a bit confused with the argument.
> Installed via pip, and not a Python?
> I think the target users are programmers just like we are.
> And hopefully, they would appreciate our efforts by reporting bugs.
> Lizard is closed-open-source agnostic when comes to analysis.
> I bet no company in the world has a proprietary dialect of C# to hide.
> The argument that the code is closed and can't report a bug is just phony.
> 
> Removing asserts just makes our life harder triaging the bugs.
> The users ignoring the asserts/bugs will risk getting invalid results.
> It just comes down to fail nosily instead of silently.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/terryyin/lizard/pull/123/files/47af1f2b3bb39f4719968a72dc38a4f41f1995a3#r65810575>, or mute the thread <https://github.com/notifications/unsubscribe/AAwJYlt7n0JT8soYgiLWOEz4sBRAUtRVks5qIjlugaJpZM4Id8An>.
>